cip/CIP2016-06-22-nested-subqueries.adoc

```diff
@@ -0,0 +1,347 @@
+
+Using a single subquery - old version using `FOREACH`:
+[source, cypher]
+----
+MATCH (r:Root)
+FOREACH(x IN range(1, 10) |
+  MERGE (c:Child {id: x})
+  MERGE (r)-[:PARENT]->(c)
+)
+----
+
+Using a single subquery - new version using `DO`:
+[source, cypher]
+----
+MATCH (r:Root)
+UNWIND range(1, 10) AS x
+DO {
+  MERGE (c:Child {id: x})
+  MERGE (r)-[:PARENT]->(c)
+}
+----
+
+Doubly-nested subquery - old version using `FOREACH`:
+[source, cypher]
+----
+MATCH (r:Root)
+FOREACH (x IN range(1, 10) |
+  CREATE (r)-[:PARENT]->(c:Child {id: x})
+  MERGE (r)-[:PUBLISHES]->(t:Topic {id: r.id + x})
+  FOREACH (y IN range(1, 10) |
+    CREATE (c)-[p:PARENT]->(:Child {id: c.id * 10 + y})
+    SET p.id = c.id * 5 + y
+  )
+)
+----
+
+Doubly-nested subquery - new version using `DO`:
+[source, cypher]
+----
+MATCH (r:Root)
+UNWIND range(1, 10) AS x AS x
+DO {
+  CREATE (r)-[:PARENT]->(c:Child {id: x})
+  MERGE (r)-[:PUBLISHES]->(t:Topic {id: r.id + x})
+  UNWIND range(1, 10) AS y
+  DO {
+    CREATE (c)-[p:PARENT]->(:Child {id: c.id * 10 + y})
+    SET p.id = c.id * 5 + y
+  }
+}
+----
+
+=== Interaction with existing features
+
+Apart from the suggested deprecation of the `FOREACH` clause, nested read-only, write-only and read-write subqueries do not interact directly with any existing features.
+
+=== Alternatives
+
+Alternative syntax has been considered during the production of this document:
+
+  * Using round braces; i.e. `MATCH (...)`
+  * Using alternative keywords:
+
+    ** `SUBQUERY`
+    ** `QUERY`
+
+== What others do
+
+=== SQL
+
+The following types of subqueries are supported in SQL:
+
+Scalar:
+[source, cypher]
+----
+SELECT orderID
+FROM Orders
+WHERE orderID =
+  (SELECT max(orderID) FROM Orders)
+----
+
+Multi-valued:
+[source, cypher]
+----
+SELECT customerID
+FROM Customers
+WHERE customerID IN
+  (SELECT customerID FROM Orders)
+----
+
+Correlated:
+[source, cypher]
+----
+SELECT orderID, customerID
+FROM Orders AS O1
+WHERE orderID =
+  (SELECT max(O2.orderID) FROM Orders AS O2
+   WHERE O2.customerID = O1.customerID)
+----
+
+Table-valued/table expression:
+[source, cypher]
+----
+SELECT orderYear
+FROM
+  (SELECT YEAR(orderDate) AS orderYear
+  FROM Orders) AS D
+----
+
+Both scalar and table expression subqueries are out of scope for the purposes of this CIP. They will be addressed in forthcoming CIPs.
+
+=== SPARQL
+
+https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#subqueries[SPARQL] only supports uncorrelated subqueries, exemplified by:
+
+[source, cypher]
+----
+SELECT ?y ?minName
+WHERE {
+  :alice :knows ?y .
+ {
+    SELECT ?y (MIN(?name) AS ?minName)
+    WHERE {
+      ?y :name ?name .
+    } GROUP BY ?y
+  }
+}
+----
+
+Owing to the bottom-up nature of SPARQL query evaluation, the subqueries are evaluated logically first, and the results are projected up to the outer query.
+
+Only variables projected out of the subquery will be visible, or in scope, to the outer query.
```

Comment:
I would tend to say that this is not a disadvantage, but rather then a (good) feature.